internal/campaigns/run_steps.go

```diff
@@ -271,3 +335,169 @@ func (e stepFailedErr) SingleLineError() string {
+		if err != nil {
+			return containerFiles, err
+		}
+
+		if err := tmpl.Execute(&out, stepCtx); err != nil {
+			return containerFiles, err
+		}
+
+		containerFiles[fileName] = &out
+	}
+
+	return containerFiles, nil
+}
+
+func renderStepEnv(env map[string]string, stepCtx *StepContext) (map[string]string, error) {
+	parsedEnv := make(map[string]string, len(env))
+
+	fnMap := stepCtx.ToFuncMap()
+
+	for k, v := range env {
+		// We treat the file contents as a template and render it
+		// into a buffer that we then mount into the code host.
+		var out bytes.Buffer
+
+		tmpl, err := template.New(k).Delims("${{", "}}").Funcs(fnMap).Parse(v)
+		if err != nil {
+			return parsedEnv, err
+		}
+
+		if err := tmpl.Execute(&out, stepCtx); err != nil {
+			return parsedEnv, err
+		}
+
+		parsedEnv[k] = out.String()
+	}
+
+	return parsedEnv, nil
+}
+
+// StepContext represents the contextual information available when executing a
+// step that's defined in a campaign spec.
+type StepContext struct {
+	PreviousStep StepResult
+	Repository   graphql.Repository
+}
+
+// ToFuncMap returns a template.FuncMap to access fields on the StepContext in a
+// text/template.
+func (stepCtx *StepContext) ToFuncMap() template.FuncMap {
+	return template.FuncMap{
+		"join": func(list []string, sep string) string {
+			return strings.Join(list, sep)
+		},
+		"previous_step": func() map[string]interface{} {
+			result := map[string]interface{}{
+				"modified_files": stepCtx.PreviousStep.ModifiedFiles(),
+				"added_files":    stepCtx.PreviousStep.AddedFiles(),
+				"deleted_files":  stepCtx.PreviousStep.DeletedFiles(),
+			}
+
+			if stepCtx.PreviousStep.Stdout != nil {
+				result["stdout"] = stepCtx.PreviousStep.Stdout.String()
+			} else {
+				result["stdout"] = ""
+			}
+
+			if stepCtx.PreviousStep.Stderr != nil {
+				result["stderr"] = stepCtx.PreviousStep.Stderr.String()
+			} else {
+				result["stderr"] = ""
+			}
+
+			return result
+		},
+		"repository": func() map[string]interface{} {
+			return map[string]interface{}{
+				"search_result_paths": stepCtx.Repository.SearchResultPaths(),
+				"name":                stepCtx.Repository.Name,
+			}
+		},
+	}
+}
+
+// StepResult represents the result of a previously executed step.
+type StepResult struct {
+	// Files are the changes made to files by the step.
+	Files StepChanges
+
+	// Stdout is the output produced by the step on standard out.
+	Stdout *bytes.Buffer
+	// Stderr is the output produced by the step on standard error.
+	Stderr *bytes.Buffer
+}
+
+// StepChanges are the changes made to files by a previous step in a repository.
+type StepChanges struct {
+	Modified []string
+	Added    []string
+	Deleted  []string
+}
+
+// ModifiedFiles returns the files modified by a step, whitespace-separated in a single string.
+func (r StepResult) ModifiedFiles() string {
+	if len(r.Files.Modified) == 0 {
+		return ""
+	}
+	return strings.Join(r.Files.Modified, " ")
```

Comment:
This pattern feels problematic: if the paths include spaces, then I presume users won't be able to distinguish a file called `a b` against files called `a` and `b`.

Fortunately, I think the solution might already be at hand: could we return `[]string` from these functions instead and require the use of the `join` function in the template?

(Additional thought: if we do this, we might want to add another function that can transform a `[]string` into a string with appropriately `sh`-quoted arguments, but that isn't on the critical path for this PR.)

Reply:
> Fortunately, I think the solution might already be at hand: could we return []string from these functions instead and require the use of the join function in the template?

Very good point! Will do that.

> (Additional thought: if we do this, we might want to add another function that can transform a []string into a string with appropriately sh-quoted arguments, but that isn't on the critical path for this PR.)

Very good idea! Something like `${{ to_shell_args repository.search_result_paths }}`